src/webgl/p5.Shader.js

```diff
@@ -0,0 +1,330 @@
+/**
+ * This module defines the p5.Shader class
+ * @module Lights, Camera
+ * @submodule Shaders
+ * @for p5
+ * @requires core
+ */
+
+var p5 = require('../core/core');
+
+/**
+ * Shader class for WEBGL Mode
+ * @class p5.Shader
+ * @constructor
+ * @param {renderer} [p5.Renderer.GL] an instance of p5.RendererGL that
+ * will provide the GL context for this new p5.Shader
+ * @param {vertSrc} [String] source code for the vertex shader (as a string)
+ * @param {fragSrc} [String] source code for the fragment shader (as a string)
+ */
+p5.Shader = function(renderer, vertSrc, fragSrc) {
+  // TODO: adapt this to not take ids, but rather,
+  // to take the source for a vertex and fragment shader
+  // to enable custom shaders at some later date
+  this._renderer = renderer;
+  this._vertSrc = vertSrc;
+  this._fragSrc = fragSrc;
+  this._vertShader = -1;
+  this._fragShader = -1;
+  this.glProgram = 0;
+  this.loadedAttributes = false;
+  this.attributes = {};
+  this.loadedUniforms = false;
+  this.uniforms = {};
+  this.bound = false;
+
+  return this;
+};
+
+/**
+ * Creates, compiles, and links the shader based on its
+ * sources for the vertex and fragment shaders (provided
+ * to the constructor). Populates known attributes and
+ * uniforms from the shader.
+ * @method init
+ */
+p5.Shader.prototype.init = function() {
+  if (this.glProgram === 0 /* or context is stale? */) {
+    var gl = this._renderer.GL;
+
+    //set up the shader by
+    // 1. creating and getting a gl id for the shader program,
+    // 2. compliling its vertex & fragment sources,
+    // 3. linking the vertex and fragment shaders
+    this._vertShader = gl.createShader(gl.VERTEX_SHADER);
+    //load in our default vertex shader
+    gl.shaderSource(this._vertShader, this._vertSrc);
+    gl.compileShader(this._vertShader);
+    // if our vertex shader failed compilation?
+    if (!gl.getShaderParameter(this._vertShader, gl.COMPILE_STATUS)) {
+      alert('Yikes! An error occurred compiling the vertex shader:' +
```

Comment:
wondering about this use of alert() here. I don't think we're using it elsewhere in the library for errors. would a friendly error suffice here? or is this one critical enough that we need to break that pattern?